### DISCUSSION TITLE
How to setError with a message type other than string

### DISCUSSION THREAD

**Discussion by rostero1** [react-hook-form/react-hook-form#5760] [Q&A]
In v7 or for v8, is there a type-safe way to set a non-string error message for a field?  

For example, let's say I'm validating a password to include a number, symbol, character, etc ...  The business owner requests some special UI to display this, where there is a checkbox behind each of the parts.  A datatype for this could be `{ number: true, symbol: false, character: true }`.  Then I can safely know that `error?.password?.raw?.number` compiles, but `error?.password?.raw?.cucumber` will not.  

I've had to return something other than a string for an error message a handful of times.  Another example could be getting a backend error and wanting to `setError('submitError', ajaxError)` or `setError('submitError', { code: 401, message: 'Bad' })` where you need an effect to run only on a 401 and you don't know what the message will be.

**Comment by jorisre** [comment]
Hi 👋🏻 

You can pass `types` to `setError`. Here is a basic instance https://codesandbox.io/s/pensive-dhawan-9ei7e
Does it help?

**Reply by rostero1** [reply]
Hi @jorisre, thanks for the reply.  Almost.  I am trying to get the following to throw an error:

`const number = errors.password?.types?.shumber;`

Instead it type checks (with the type being ValidateResult).

I couldn't find information about the `types` field on the web docs for `useForm` `setError`, so thanks for the hint.

**Reply by jorisre** [reply]
You can find a similar example in the documentation, it's the 3rd codesanbox instance: https://react-hook-form.com/api/useform/seterror

Do you still have the issue?

**Reply by rostero1** [reply]
Thanks again.  I still have an issue because the following compiles:

https://codesandbox.io/s/delicate-fast-yszmz?file=/src/App.tsx

```
{errors.username && <p>{errors.username?.types?.OOPS}</p>}
{errors.username && <p>{errors.username?.types?.BAD}</p>}
```

```
export type MultipleFieldErrors = {
  [K in keyof RegisterOptions]?: ValidateResult;
} & {
  [key: string]: ValidateResult;
};
```

**Reply by jorisre** [reply]
I'm not sure to understand what you are trying to achieve

**Reply by rostero1** [reply]
Utilize typescript to achieve type-safe form code.  `useForm` has a type parameter for the form values.  I was wondering in the future if there could be one on the form errors.

**Comment by Noitidart** [comment]
I think your question is like mine, could you please check this question - https://github.com/react-hook-form/react-hook-form/discussions/8401

**Reply by Moshyfawn** [reply]
I've answered your question [here](https://github.com/react-hook-form/react-hook-form/discussions/8401#discussioncomment-2819633)

**Reply by Noitidart** [reply]
Thank you very very much sir!
